spark/v4.0/spark/src/main/java/org/apache/iceberg/spark/source/ExtractRowLineage.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.spark.source;
+
+import java.util.List;
+import java.util.function.Function;
+import org.apache.iceberg.MetadataColumns;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableList;
+import org.apache.spark.sql.catalyst.InternalRow;
+import org.apache.spark.sql.catalyst.ProjectingInternalRow;
+import org.apache.spark.sql.catalyst.expressions.GenericInternalRow;
+import org.apache.spark.sql.types.LongType$;
+import org.apache.spark.sql.types.StructType;
+import scala.collection.JavaConverters;
+
+class ExtractRowLineage implements Function<InternalRow, InternalRow> {
+  private static final StructType ROW_LINEAGE_SCHEMA =
+      new StructType()
+          .add(MetadataColumns.ROW_ID.name(), LongType$.MODULE$, true)
+          .add(MetadataColumns.LAST_UPDATED_SEQUENCE_NUMBER.name(), LongType$.MODULE$, true);
+  private final Schema writeSchema;
+
+  private ProjectingInternalRow cachedRowLineageProjection;
+
+  ExtractRowLineage(Schema writeSchema) {
+    this.writeSchema = writeSchema;
+  }
+
+  @Override
+  public InternalRow apply(InternalRow meta) {
+    // If output schema is null, i.e. deletes in MoR, or row lineage is not required on write,
+    // return a null row
+    if (writeSchema == null || writeSchema.findField(MetadataColumns.ROW_ID.name()) == null) {
```

Comment:
I thought we need lazy initialization when `writeSchema` is null for the cases described in the comment. I assume non-null `meta` would contain the write schema from `ProjectingInternalRow`.

if we only need this if-check, it can be moved to the constructor only.

do we also need to distinguish between the two cases of `writeSchema` is null vs doesn't contain `ROW_ID`? the later is where row_lineage is not enabled (like V2) table?

Reply:
So in the callers of `createWriter` method from `SparkWrite` and `SparkPositionDeltaWrite`, the `writeSchema` won't be null?

from tracing the code, this seems correct to me.

Reply:
Correct. I restructured the writer initialization so now we're guaranteed that `ExtractRowLineage` will always be passed a non-null `writeSchema`